src/input/mod.rs

```diff
@@ -265,27 +272,35 @@ impl State {
     where
         I::Device: 'static,
     {
-        let device_output = event.device().output(self);
+        let device = event.device();
+        let device_output = device.output(self);
         let device_output = device_output.as_ref();
-        let (target_geo, keep_ratio, px, transform) =
-            if let Some(output) = device_output.or_else(|| self.niri.output_for_tablet()) {
-                (
-                    self.niri.global_space.output_geometry(output).unwrap(),
-                    true,
-                    1. / output.current_scale().fractional_scale(),
-                    output.current_transform(),
-                )
-            } else {
-                let geo = self.global_bounding_rectangle()?;
 
-                // FIXME: this 1 px size should ideally somehow be computed for the rightmost output
-                // corresponding to the position on the right when clamping.
-                let output = self.niri.global_space.outputs().next().unwrap();
-                let scale = output.current_scale().fractional_scale();
+        let device = (&device as &dyn Any).downcast_ref::<input::Device>();
+
+        let (target_geo, keep_ratio, px, transform) = if let Some(output) =
+            device_output.or_else(|| {
+                device
+                    .and_then(|device| self.niri.devices.get(device))
+                    .and_then(|data| self.niri.output_for_tablet(&data.name))
```

Comment:
Doesn't this change the logic slightly? Before, `output_for_tablet()` would be used for all devices, now only for devices that can downcast to `input::Device`.